aws/ratelimit/token_bucket.go

```diff
@@ -0,0 +1,53 @@
+package ratelimit
+
+import (
+	"sync"
+)
+
+// TokenBucket provides a concurrency safe utility for adding and removing
+// tokens from the available token bucket.
+type TokenBucket struct {
+	capacity    uint
+	maxCapacity uint
+
+	// TODO would it be better to replace this mutex with CAS loops?
+	mu sync.Mutex
+}
+
+// NewTokenBucket returns an initialized TokenBucket with the capacity
+// specified.
+func NewTokenBucket(i uint) *TokenBucket {
+	return &TokenBucket{
+		capacity:    i,
+		maxCapacity: i,
+	}
+}
+
+// Retrieve attempts to reduce the available tokens by the amount requested. If
+// available tokens are already negative, or the retrieve would make available
```

Comment:
np: Wording that available tokens are already negative, technically they can't be negative since the type is uint.